### ISSUE TITLE
New-UDGrid - Errors when server in a foreach is unavailable

### ISSUE THREAD

**Issue by dtaboada** [ironmansoftware/universal-dashboard#14]
Hey guys, I'm using new-udgrid to output a list of data Ive put together via pscustomobject, one of them being test-connection to a list of servers,(IsAlive/Dead states in a foreach) However when ever I test this and disable one of the servers nics to see if the ping data would reflect on the grid, the component throws a  Testing connection to computer 'Server000' failed: Error due to lack of resources. Any ideas? 

Would appreciate any feedback, open to any suggestions.

Thanks,
DT

**Comment by dtaboada** [comment]
I introduced try catch to trap the error and set it to something, that appears to have worked when all machines are ON, the script/UD have no issue but when testing(restarting or disabling one of the servers NICs in the list) it fails with the index error.  Below is a codesnip, what am I missing?

I went ahead and enabled logging for UD and this is being recorded (22:35:06 ComponentController {"Error":{"message":"Index was out of range. Must be non-negative and less than the size of the collection.\r\nParameter name: index","location":"   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)\r\n   at System.Collections.Generic.List`1.get_Item(Int32 index)\r\n   at System.Management.Automation.PSDataCollection`1.get_Item(Int32 index)\r\n   at PowerShellProTools.UniversalDashboard.Execution.ExecutionService.ExecuteEndpoint(ExecutionContext context)","type":"error","id":null,"refreshInterval":0,"autoRefresh":false}})

New-UDCard -Title "Title Basic Ping Tests" -Watermark check
                New-UDGrid -DebugEndpoint  -Title "Basic Ping Tests" -Headers @("Computername","Ping","Status","ResponseTime","Role") -Properties @("Computername","Ping","Status","ResponseTime","Role") -AutoRefresh -RefreshInterval 5 @Colors -PageSize 30  -Endpoint {
                    $Output += foreach ($srv in $webservers) {
                    
                    try {
                        $Pinger = Test-Connection -ComputerName $srv -Count 1 -ErrorAction SilentlyContinue -ErrorVariable pingu |out-null 
                        
                                $props = @{
                
                                        Computername = $srv
                                        Status         = "ONLINE"
                                        }
                                        New-Object -TypeName PSObject -Property $Props
                              }catch {
                                    try {
                                        Test-Connection -ComputerName $srv -Count 1 -ErrorAction SilentlyContinue -ErrorVariable pingu |out-null 
                                        $Props = @{
                                            ComputerName   = $srv
                                            Status = $(if ($pingu) {"PROBLEM"})
                                            }
                                        New-Object -TypeName PSObject -Property $Props
                                    } catch {
                                        $Props = @{
                                            ComputerName   = $srv
                                            Status         = "No response to ping"
                                            }
                                        New-Object -TypeName PSObject -Property $Props
                                    }
                                    }   
                            }
                            $Output |Out-UDGridData

**Comment by adamdriscoll** [comment]
This sounds like an internal error to UD when nothing is returned by Out-UDGridData or something like that. I will take a look.

**Comment by dtaboada** [comment]
Great! Much appreciated...

**Comment by Crowbar3** [comment]
I know you marked this as a bug already, just wanted to give my input in case it helps...

It seems that the Try/Catch commands don't work so well because you usually put in commands that inherently fail.  So things like test-connection where a server is offline or get-adcomputer where the object doesn't exist result in the "index was out of range" error.

> {"error":{"message":"Index was out of range. Must be non-negative and less than the size of the collection.\r\nParameter name: index","location":"   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)\r\n   at System.Collections.Generic.List`1.get_Item(Int32 index)\r\n   at System.Management.Automation.PSDataCollection`1.get_Item(Int32 index)\r\n   at PowerShellProTools.UniversalDashboard.Execution.ExecutionService.ExecuteEndpoint(ExecutionContext context)","type":"error","id":null,"refreshInterval":0,"autoRefresh":false,"hasCallback":false}}

For Test-Connection I ended up using the -quiet command to generate a True or False, so that could be a work-around for dtaboada for the time being.

```
                If (Test-Connection -ComputerName $servername -Count 1 -Quiet) {
                    New-UDCard -Title "Server is online!" -BackgroundColor darkgreen -FontColor white -Text $servername
                    } Else {
                    New-UDCard -Title "Server is offline!" -BackgroundColor darkred -FontColor white -Text $servername
                    }
```

I would still love to see the Try/Catch commands work as intended at some point.  Been loving UD so far, very easy to use.  Thanks!

**Comment by adamdriscoll** [comment]
Fixed in 1.5.2
